### PR TITLE
[MOD-2845] fix: prev. inadequate handling of stale in-mem clients

### PR DESCRIPTION
**Pull request by thundergolfer**
## Describe your changes

- MOD-2845

Got a local reproduction of some stale creds. Fix locally by clearing the Client's channel which was initialized with the old creds and left in-memory. In retrospect, this is an obvious problem 😖. 

<details> <summary>Backward/forward compatibility checks</summary>

- [x] Client+Server: this change is compatible with old servers
- [x] Client forward compatibility: this change ensures client can accept data intended for later versions of itself

---

</details>

### CODE REVIEW

File: test/container_app_test.py [hunk 1]
```diff
@@ -73,6 +73,34 @@ async def test_container_snapshot_restore(container_client, tmpdir, servicer):
...
+    f = Function.lookup(app_name, "square", client=container_client)
+    await f.remote.aio()
+    print(servicer.function_map_metadata)
+    io_manager = ContainerIOManager(api_pb2.ContainerArguments(), container_client)
+    restore_path = temp_restore_path(tmpdir)
+    with mock.patch.dict(
+        os.environ, {"MODAL_RESTORE_STATE_PATH": str(restore_path), "MODAL_SERVER_URL": servicer.container_addr}
+    ):
+        io_manager.memory_snapshot()
+        # RPC should have use credentials, not old credentials
```
Comment (thundergolfer) [inline test/container_app_test.py:98]:
```suggestion
        # RPC should have used new credentials, not old credentials
```

File: modal/client.py [hunk 2]
```diff
@@ -116,14 +121,14 @@ def authenticated(self) -> bool:
...
     def credentials(self) -> tuple:
         """mdmd:hidden"""
-        if self._credentials is None and self.client_type == api_pb2.CLIENT_TYPE_CONTAINER:
+        if self._credentials is None and self._snapshotted:
             logger.debug("restoring credentials for memory snapshotted client instance")
             self._credentials = (config["task_id"], config["task_secret"])
         return self._credentials
 
-    async def _open(self):
+    def _open(self):
```
Comment (thundergolfer) [inline modal/client.py:129]: Seemed this didn't need to be `async`?

File: test/container_app_test.py [hunk 3]
```diff
@@ -75,7 +75,24 @@ async def test_container_snapshot_restore(container_client, tmpdir, servicer):
...
+    from modal import Function
+    from modal.runner import deploy_app
+
+    app.function()(square)
+    app_name = "my-app"
+    deploy_app(app, app_name, client=container_client).app_id
+    f = Function.lookup(app_name, "square", client=container_client)
+    await f.remote.aio()
+    # RPC should have used new credentials, not old credentials
+    await f.remote.aio()
```
Comment (erikbern) [inline test/container_app_test.py:93]: i don't fully understand this test. where do you simulate the actual snapshot+restore?

**Review by erikbern — APPROVED** [review]
lgtm!

if we inject creds worker side then i think we can remove this code?

**Comment by thundergolfer** [comment]
> if we inject creds worker side then i think we can remove this code?

we do actually do this already. We swap out the creds and emit a warning.

**Comment by cweld510** [comment]
@thundergolfer does this need to be merged still?

**Pull request closed without merging**